---
- name: Verify
  hosts: localhost
  gather_facts: false
  vars:
    project_dir: "{{ lookup('env', 'MOLECULE_PROJECT_DIRECTORY') }}"
    dtas_dir: "{{ lookup('env', 'MOLECULE_PROJECT_DIRECTORY') }}/dtas"
    test_config_file: "{{ lookup('env', 'MOLECULE_PROJECT_DIRECTORY') }}/{{ lookup('env', 'MOLECULE_IT_TEST_CONFIG') }}"
  tasks:

    - name: Gather EC2 facts
      delegate_to: "{{ groups['nginx'][0] }}"
      amazon.aws.ec2_metadata_facts:
      ignore_errors: yes

    - name: Install pytest requirements
      pip:
        chdir: "{{ dtas_dir }}"
        requirements: requirements.txt

    - name: Configure deployment URL in {{ test_config_file }}
      ansible.builtin.replace:
        path: "{{ test_config_file }}"
        regexp: 'TEST_URL'
        replace: "https://{{ ansible_ec2_public_hostname | default('localhost') }}"

    - name: Run tests
      block:
        - name: Run pytest
          delegate_to: localhost
          vars:
              selfsigned_cert_path: "{{ project_dir }}/configuration_files/ssl_certificates/{{ ansible_ec2_public_hostname | default('localhost') }}.crt"
          ansible.builtin.shell:
            chdir: "{{ dtas_dir }}"
            cmd: |
              export REQUESTS_CA_BUNDLE={{ selfsigned_cert_path }}
              pytest --tb=line --configuration {{ test_config_file }} tests/ -s
      rescue:
        - name: Print multiline pytest stdout as best as we can
          debug:
            msg: "{{ ansible_failed_result.stdout_lines }}"
        - name: Explicit failure after printing failure debug
          fail: msg="Aborting due to pytest failure"
